docs/design/2024-05-10-extension-authentication-plugin.md

```diff
@@ -0,0 +1,231 @@
+
+Following is the interface for an auth plugin:
+```go
+// AuthPlugin is an interface for authentication plugins.
+type AuthPlugin interface {
+    // Name returns the name of the auth plugin. It will be registered as a system variable in TiDB which can be used inside the `CREATE USER ... IDENTIFIED WITH 'plugin_name'` statement.
+    Name() string
+    
+    // PasswordReader is a function to communicate with the client during handshake phase to 
+    // read/write more data to the client-side plugin using authConn. For example, sending extra salt to the client.
+    // If `plugin.PasswordReader()` returns nil, the connection will default to mysql_clear_password plugin which means the password will be passed from the client-side in clear text format. Otherwise, the connection will use `plugin.Name()` as the current plugin
+    // When `plugin.PasswordReader()(ctx, authConn, resp)` is called, it should return the final password received from the client in bytes format. This bytes password will be passed into `plugin.Authenticate` below.
+    // ctx: context passed in from caller
+    // authConn: interface for the plugin to communicate with the client
+    PasswordReader() func(ctx context.Context, authConn conn.AuthConn) ([]byte, error)
+    
+    // Authenticate is called when a client connects to the server as a user and the server authenticates the user.
+    // If an error is returned, the login attempt fails, otherwise it succeeds.
+    // authUser: the username in the connect attempt
+    // storedPwd: the user's password stored in mysql.user table
+    // inputPwd (authentication): the user's password passed in from the connection attempt in bytes (return value of `plugin.PasswordReader()(ctx, authConn, resp)` if `plugin.PasswordReader() != nil`)
+    // salt: randomly generated salt for the current connection
+    // connState: the TLS connection state (contains the TLS certificate) if client is using TLS. It will be nil if the client is not using TLS
+    // authConn: interface for the plugin to communicate with the client
+    Authenticate(authUser string, storedPwd string, inputPwd []byte, salt []byte, connState *tls.ConnectionState, authConn conn.AuthConn) error
+    
+    // EncodePassword is a function for user to implement customized ways to encode the password (e.g. hash/salt/clear-text). The returned string will be stored as the encoded password in the mysql.user table.
+    // If the input password is considered as invalid, this should return an error.
+    // pwd: User's input password in CREATE/ALTER USER statements in clear-text
+    EncodePassword(pwd string) (string, bool)
+    
+    // IsValidPasswordHash checks if the password hash stored in the mysql.user table is valid.
+    // This is called when retrieving an existing user to make sure the password stored is valid and not modified.
+    // pwdHash: hash of the password stored in the internal user table
+    IsValidPasswordHash(pwdHash string) bool
```

Comment:
Are `GenerateAuthString`/`ValidateAuthString` better names? Because mysql use `generate_authentication_string ` and `validate_authentication_string` as the function names: https://dev.mysql.com/doc/extending-mysql/8.0/en/writing-authentication-plugins-server-side.html

Reply:
`EncodePassword` and `IsValidPasswordHash` are picked to be in-sync with current naming in TiDB, but yes I think it's better to match MySQL naming as close as possible. Will update the doc.